Chia-Network/chia-blockchain
# remote and browser based UI

This PR enables the react UI to interact with deamons running on remote machines. This allows the farmer/wallet/full node to be run headless on a pi, but still be graphically managed via the react UI.

With these changes the UI can be run via electron and also as a node module with a browser based UI. In either mode, the UI can interact with a daemon on the `localhost` or remotely (with `localhost` always being the default). Whenever running in the browser, or with a remote daemon, the UI will not attempt to start or stop the daemon. When running via electron, on the same machine as the daemon, it will start and stop along with the UI as it currently does.

To test the node app
````
cd electron-react
npm install
npm start
````
A browser will open at `localhost:3000` to host the UI. To attach the browser UI to a remote daemon, the remote host name can be supplied as a query parameter: `localhost:3000?selfHostName=<remote machine name or ip>`

To run the electron UI against a remote daemon, the remote host name can be passed via the command line: `Chia.exe --selfHostName=<remote machine name or ip>`

In all cases the UI expects the daemon to be listening on port 55400. 

To run the UI from a browser, the _nodejs_ service needs to be running somewhere on the network. Depending on where you run the _nodejs_ service you may need to change the port the daemon listens to:
- If the _nodejs_ service and dameon are not on the same machine, `config.yaml` needs to have the daemon listen on `0.0.0.0` instead of `localhost`. Find the unindented line `self_hostname: localhost` and change it to `self_hostname: 0.0.0.0`. _note:_ doing this intentionally exposes the daemon on the network, so firewall rules should be such that traffic is restricted to only the UI hosts you intend. With `ufw` for instance add a rule like this `ufw allow from <ip of ui machine> to any port 55400 proto tcp`
- Alternatively the _nodejs_ service can run on that same server as the daemon, in which case the service can talk to the daemon on `localhost `but will need a [web server to host the service](https://pm2.keymetrics.io/docs/usage/quick-start/). This is a bit more complicated to setup, but has advantages for simplicity of use and security. (you'd still want to use firewall or webserver rules to control client access as the UI does not authenticate.)
